src/keygen/Main.hs

```diff
@@ -1,121 +1,87 @@
-
-main :: IO ()
-main = do
-    KO {..} <- execParser optsInfo
-    let keysDir = takeDirectory koPattern
-        genFileDir = takeDirectory koGenesisFile
+getTestnetGenesis :: Coin -> TestStakeOptions -> IO GenesisData
+getTestnetGenesis avvmStake tso@TSO{..} = do
+    let keysDir = takeDirectory tsoPattern
     createDirectoryIfMissing True keysDir
-    createDirectoryIfMissing True genFileDir
 
-    genesisList <- forM [1..koStakeholders] $ \i ->
-        generateKeyfile $ replace "{}" (show i) koPattern
-    print $ show koStakeholders ++ " keyfiles are generated"
+    let totalStakeholders = tsoRichmen + tsoPoors
+    genesisList <- forM [1 .. totalStakeholders] $ \i ->
+        generateKeyfile $ replace "{}" (show i) tsoPattern
+    print $ show totalStakeholders ++ " keyfiles are generated"
 
-    let distr = TestnetStakes
-            { sdTotalStake = mkCoin koTotalStake
-            , sdRichmen    = koRichmen
-            , sdPoor       = koStakeholders - koRichmen
-            }
+    let distr = genTestnetStakes avvmStake tso
         genesisAddrs = map (makePubKeyAddress . fst) genesisList
         genesisVssCerts = HM.fromList
                           $ map (_1 %~ addressHash)
-                          $ take (fromIntegral koRichmen) genesisList
+                          $ take (fromIntegral tsoRichmen) genesisList
         genData = GenesisData
             { gdAddresses = genesisAddrs
             , gdDistribution = distr
             , gdVssCertificates = genesisVssCerts
             }
-    BSL.writeFile koGenesisFile $ encode genData
-    case decodeFull (encode genData) of
-        Right (_ :: GenesisData) ->
+    return genData
+
+getAvvmGenesis :: AvvmStakeOptions -> IO GenesisData
+getAvvmGenesis ASO{..} = do
+    jsonfile <- BSL.readFile asoJsonPath
+    holder <- getHolderId asoHolderKeyfile
+    case eitherDecode jsonfile of
+        Left err       -> error $ toText err
+        Right avvmData -> pure $
+            genGenesis avvmData asoIsRandcerts holder
+
+main :: IO ()
+main = do
+    KO {..} <- execParser optsInfo
+    let genFileDir = takeDirectory koGenesisFile
+    createDirectoryIfMissing True genFileDir
+
+    mAvvmGenesis <- traverse getAvvmGenesis koAvvmStake
+    let avvmStake = maybe (mkCoin 0)
+            (getTotalStake . gdDistribution) mAvvmGenesis
+    mTestnetGenesis <- traverse (getTestnetGenesis avvmStake) koTestStake
+
+    let mGenData = mappend <$> mTestnetGenesis <*> mAvvmGenesis
+                   <|> mTestnetGenesis
+                   <|> mAvvmGenesis
+        genData = maybe (error "At least one of options \
+                               \(AVVM stake or testnet stake) \
+                               \should be provided")
+                  identity mGenData
+        binGenesis = encode genData
+
+    case decodeFull binGenesis of
+        Right (_ :: GenesisData) -> do
             putText "genesis.bin generated successfully\n"
+            BSL.writeFile koGenesisFile binGenesis
```

Comment:
I see no significant reason to choose between lazy and non-lazy bs, so let's consider the fact that `encode` is shorter than `encodeStrict` as a reason.